addons/account/models/account_account.py

```diff
@@ -134,29 +135,55 @@ def _check_account_type_unique_current_year_earning(self):
     def _field_to_sql(self, alias: str, fname: str, query: (Query | None) = None, flush: bool = True) -> SQL:
         if fname == 'internal_group':
             return SQL("split_part(account_account.account_type, '_', 1)", to_flush=self._fields['account_type'])
+        if fname == 'code':
+            field = self._fields.get(fname)
+
+            company_dependent_field_alias = query.make_alias(alias, f'{field.name}_company_dependent')
+            query.add_join('LEFT JOIN', alias=company_dependent_field_alias, table='ir_property', condition=SQL(
+                    """
+                        %(ir_property_fields_id)s = %(field_id)s
+                        AND %(ir_property_company_id)s = %(root_company_id)s
+                        AND %(ir_property_res_id)s = 'account.account,' || %(account_id)s::text
+                    """,
+                    ir_property_fields_id=self.env['ir.property']._field_to_sql(company_dependent_field_alias, 'fields_id'),
+                    field_id=self.env['ir.model.fields']._get(self._name, field.name).id,
+                    ir_property_company_id=self.env['ir.property']._field_to_sql(company_dependent_field_alias, 'company_id'),
+                    root_company_id=self.env.company.root_id.id,
+                    ir_property_res_id=self.env['ir.property']._field_to_sql(company_dependent_field_alias, 'res_id'),
+                    account_id=SQL.identifier(alias, 'id'),
+                )
+            )
+
+            # This is the field on ir.property that will contain the value of 'code'.
+            value_field = self.env['ir.property']._fields[TYPE2FIELD[field.type]]
+
+            return self.env['ir.property']._field_to_sql(company_dependent_field_alias, value_field.name)
+
         return super()._field_to_sql(alias, fname, query, flush)
 
-    @api.constrains('company_id', 'code')
+    @api.constrains('company_ids', 'code')
     def _constrains_code(self):
-        # check for duplicates in each root company
-        by_root_company = self.grouped(lambda record: record.company_id.root_id)
-        for root_company, records in by_root_company.items():
-            by_code = records.grouped('code')
-            if len(by_code) < len(records):
-                # retrieve duplicates within self
-                duplicates = next(recs for recs in by_code.values() if len(recs) > 1)
-            else:
-                # search for duplicates of self in database
-                duplicates = self.search([
-                    ('company_id', 'child_of', root_company.id),
```

Comment:
You need to keep the company in the domain because `with_company` doesn't remove other companies, it only reorders them.
So the `search` would still return results from other companies.